### ISSUE TITLE
Add ablility to use variables in url path

### ISSUE THREAD

**Issue by mlamutt** [prometheus-msteams/prometheus-msteams#36]
Hi team, 
First off, thanks for this app.  I would like to add a feature request to allow for variables to be used in the URL.  We have a use case where we have dozens of MS team channels that we are trying to set up, and it would very useful to be able to set up a single receiver that uses a variable to select the correct url path.  In the following example I am adding annotations for email and the msteams connector.  It would be useful to be able to use the {{ .annotation.msteams }} variable in the receiver so I do not have to have dozens of receivers, with each connector in the url.

rule:
alert: Alert
  testing rule new alert
expr: up{hostname="xxxx.dev.local",job="deploydirector"}
  == 1
for: 1m
labels:
  channel: msteams,email,monolith
  severity: Critical
annotations:
  application: git.tools.foo.net
  description: '{{ $labels.job }} server has been down for more than 5 minutes.'
  documentation_link: https://confluence.com/display/MOD/Deployments+FAQ
  email: operations@foo.com
  msteams: Operations
  summary: THIS IS A TEST- {{ $labels.job }} is down- THIS IS A TEST

alertmanager config:
global:
  resolve_timeout: 5m

route:
  receiver: email
  group_by:
  - alertname
  routes:
  - receiver: email
    match_re:
      channel: ^(?:.*(email).*)$
    continue: true
  - receiver: msteams
    match_re:
      channel: ^(?:.*(msteams).*)$
    continue: true
  - receiver: monolith
    match_re:
      channel: ^(?:.*(monolith).*)$
    repeat_interval: 8737h
  group_wait: 30s
  group_interval: 1m
  repeat_interval: 4h
receivers:
- name: email
  email_configs:
  - send_resolved: true
    to: '{{ .CommonAnnotations.email }}'
    from: alertmanager@alerts..com
    hello: localhost
         To: '{{ .CommonAnnotations.email }}'
    html: '{{ template "email.default.html" . }}'
    require_tls: false

- name: msteams
  webhook_configs:
  - send_resolved: true
    http_config: {}
    url: "http://localhost:2000/{{ .CommonAnnotations.msteams }}'

- name: monolith
  webhook_configs:
  - send_resolved: true
    http_config:
      tls_config:
        insecure_skip_verify: true
    url: https://portal.local/api/Monolith/SendAlertsDataProd_RAW
templates: []

Would this be possible?
thanks

**Comment by Knappek** [comment]
Hi @mlamutt, I don't think this is possible as the prometheus-msteams webserver has to provide an endpoint which the alertmanager can use. You can't expose endpoints dynamically. 
It is possible though that you deploy the alertmanager together with the msteams webserver and parameterize it during the deployment. 
@bzon what are your thoughts?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by mlamutt** [comment]
Hi @Knappek,   I am not clear on what is going on under the hood.  I am setting each endpoint in the prometheus-msteams config file manually, so the list of available endpoints should be there. I am asking for the webhook url to accept a variable, so we can use:

url: "http://localhost:2000/{{ .CommonAnnotations.msteams }}'
where ({ .CommonAnnotations.msteams }} = team1 or team2... teamX

instead of 

url: "http://localhost:2000/team1'
url: "http://localhost:2000/team2'
etc,,,

In both cases, team1 and team2 have existing endpoints setup in the config.yml fuel as existing connectors.

Thanks

**Comment by AndreasKnapp1812** [comment]
@mlamutt I don't know if I understand you correctly. The config file where you define the `url` parameter is the config file of the prometheus alertmanager. I think there is already an issue for exactly what you've described: https://github.com/prometheus/alertmanager/issues/684 . 

Does it help you?

**Comment by Knappek** [comment]
@mlamutt I don't know if I understand you correctly. The config file where you define the url parameter is the config file of the prometheus alertmanager. I think there is already an issue for exactly what you've described: prometheus/alertmanager#684 .

Does it help you?

**Comment by mlamutt** [comment]
@Knappek this is the issue that I am looking at.  In a different thread (https://groups.google.com/forum/#!msg/prometheus-users/Bg0eFP_Rd9Q/8CNT05zPBwAJ), it was suggested that I reach out to you to the help.  

My goals is to be able to use a single Alertmanager Receiver in the config file instead of having to insert dozens of routes and receivers

example:
route:
  - match_re:
      channel: msteams1
    receiver: msteams1
    continue: true

  - match_re:
      channel: msteams2
    receiver: msteams2
    continue: true

  - match_re:
      channel: msteamsN
    receiver: msteamsN
    continue: true

receivers:
- name: msteams1
  webhook_configs:
    - url: "http://localhost:2000/msteams1"  
      send_resolved: true

- name: msteams2
  webhook_configs:
    - url: "http://localhost:2000/msteams2"  
      send_resolved: true

- name: msteamsN
  webhook_configs:
    - url: "http://localhost:2000/msteamsN"  
      send_resolved: true

Using variables I could shrink this to simple config where {{.GroupLabels.msteams}} is a label that is passed in.

route:
  - match_re:
      channel: .*(msteams).*
    receiver: msteams
    continue: true

receivers:
- name: msteams
  webhook_configs:
    - url: "http://localhost:2000/{{.GroupLabels.msteams}}"  
      send_resolved: true

I hope I am explaining what my request is?  thanks for looking at this.

**Comment by Knappek** [comment]
@mlamutt Now I get your point. Hmm, I think theoretically it could be possible and might be an enhancement, especially if you have lots of teams channels. 
It could be quite tricky though to achieve backwards compatibility. Either you have to expose the root `/` endpoint that is able to distribute the requests based on variables or maybe using the `?` operator where the url to prometheus-msteams looks like http://localhost:2000/dynamicEndpoint?channel={{.GroupLabels.msteams}}. 
Other suggestions are welcome..

**Comment by bzon** [comment]
@mlamutt I want to understand the problem more. 

Is this what you are currently doing?

alert manager settings:

```yaml
receivers:
- name: 'msteams1'
  webhook_configs:
    - send_resolved: true
      url: 'http://localhost:2000/{{ .CommonAnnotations.msteams1 }}'
- name: 'msteams2'
  webhook_configs:
    - send_resolved: true
      url: 'http://localhost:2000/{{ .CommonAnnotations.msteams2 }}
```

promteams connector settings:

```yaml
connectors:
- WHATEVER_THE_VALUE_OF_MSTEAMS_1: "https://outlook.office.com/webhook/xxxx/aaa/bbb"
- WHATEVER_THE_VALUE_OF_MSTEAMS_2: "https://outlook.office.com/webhook/xxxx/aaa/ccc"
```

What is the example set of configuration do you want to use? Can you provide an example user experience? 😄

**Comment by jangrewe** [comment]
I've been banging my head against the wall because of the same issue, too many channels with too many webhooks...
The big problem: the `url` field of `webhook_configs` is just a `<string>` field that doesn't allow templating - and the Prometheus team refuses to change this (there's an issue in the AM repo about this).

The solution i'm working on: using labels in the alerts to set the desired endpoint name, and parsing the alerts when received on a specific endpoint, to rewrite the request so that they look to prometheus-msteams like they were sent to a particular other endpoint.

I have the basics done, just struggling a bit with the peculiarities of Go (e.g. clearing out the request body after decoding it, losing the whole body for the actual alerting).

**Comment by bzon** [comment]
@jangrewe 

It’s been a while since this issue was opened. I’m actually quite confused now with what we want to achieve now.

Is this a feature request to dynamically choose the microsoft teams url to use based on the prometheus alert labels or annotations?

**Comment by jangrewe** [comment]
Yes, for me/us, it's about being able to dynamically route alerts to specific Teams channels based on alert metadata (labels), without having to manage all the (Teams) webhook URLs via separate receivers - which is not possible in Alertmanager itself, as it doesn't allow templating in the URL field for a webhook_config.
I'll send you a PR when it's in a usable state...

**Comment by jangrewe** [comment]
Here's a first proposal how this could be handled: https://github.com/bzon/prometheus-msteams/pull/83
I'm happy to discuss how this could be done better, as i'm not even remotely a Go dev ;-)

**Comment by kbudde** [comment]
Unfortunately the change in #83 would need to be adjusted to fit current code base. 
It would be nice if someone can invest the time to implement this (specifying a label to the choose the right webhook_config).

On the other hand there is already the option to pass the teams webhook url from alertmanager (using `/_dynamicwebhook/`).
This still needs one receiver for every team but it must only be configured in alertmanager configuration which simplifies the process a lot.  This also allows to use the alertmanager grouping and hierarchy functions which is probably a good solution.
